File: 00_software_engineering_foundations/29_tictactoe_project/game.rb
require_relative 'board'
require_relative 'human_player'
require_relative 'computer_player'

class Game
    def initialize(board_size, players)
        @players = players.map { |mark, computer| computer ? ComputerPlayer.new(mark) : HumanPlayer.new(mark) }
        @board = Board.new(board_size)
    end

    def current_player
        @players.first
    end

    def switch_turn
        @players.rotate!
        @players.first
    end

    def play
        while @board.empty_positions?
            @board.print

            pos = current_player.get_position(@board.legal_positions)
            @board.place_mark(pos, current_player.mark)

            if @board.win?(current_player.mark)
                puts "Player #{current_player.mark} won!"
                @board.print
                return
            end

            switch_turn
        end

        puts "It's a draw!"
        @board.print
    end
end

File: 04_javascript/07_tic_tac_toe_js_edition/game.js
const Board = require('./board.js');
const HumanPlayer = require('./human_player.js');
const AIPlayer = require('./ai_player.js');

class Game {
  constructor(reader) {
    this.reader = reader;
    this.board = new Board();
    this.players = [
      new HumanPlayer(this.board, 'X'),
      new AIPlayer(this.board, 'O'),
    ];
  }

  swapTurn() {
    this.players.push(this.players.shift());
  }

  run(completionCallback) {
    this.board.print();

    if (this.board.won()) {
      console.log('Winner was player ', this.board.winner());
      completionCallback();
    } else if (this.board.full()) {
      console.log("It's a draw!");
      completionCallback();
    } else {
      this.players[0].promptMove(this.reader, (pos) => {
        if (this.board.placeMark(pos, this.players[0].mark)) {
          this.swapTurn();
        } else {
          console.log('Invalid move!');
        }
        this.run(completionCallback);
      });
    }
  }
}

module.exports = Game;
